common/src/main/java/com/wynntils/models/profession/ProfessionModel.java

```diff
@@ -92,18 +103,27 @@ public void resetValueFromItem(ItemStack professionInfoItem) {
         professionProgressMap = levels;
     }
 
-    public void updateValue(ProfessionType type, float newPercentage, float xpGain) {
+    public void updatePercentage(ProfessionType type, float newPercentage, float xpGain) {
         ProfessionProgress oldValue = professionProgressMap.getOrDefault(type, ProfessionProgress.NO_PROGRESS);
 
-        // Assume a level up if the progress is less than the previous value.
-        if (oldValue.progress() > newPercentage) {
-            professionProgressMap.put(type, new ProfessionProgress(oldValue.level() + 1, newPercentage));
-        } else {
-            professionProgressMap.put(type, new ProfessionProgress(oldValue.level(), newPercentage));
+        // We leveled up, but we don't know how many times.
+        // Set the progress, level will be parsed from other messages.
+        if (newPercentage == 100) {
+            newPercentage = 0;
         }
+
+        professionProgressMap.put(type, new ProfessionProgress(oldValue.level(), newPercentage));
+
         rawXpGainInLastMinute.get(type).put(xpGain);
     }
 
+    public void updateLevel(ProfessionType type, int newLevel) {
+        ProfessionProgress oldValue = professionProgressMap.getOrDefault(type, ProfessionProgress.NO_PROGRESS);
+
+        // We don't know the progress, so we set it to 0.
```

Comment:
Is it not better to not touch the progress map here? It will be set by updatePercentage, yes?

Reply:
No sadly, become we get a level up message first, then a +1000xp [100%]. (So we can only set this to 0 in both cases of level and progress update).

Reply:
But I guess this is not required here yes.